[CIR][CIRGen] Support OpenCL Vector Types

seven-mile opened the pull request:
Resolve #532 .

Support CIRGen of `ExtVectorElementExpr` that includes swizzle `v.xyx` and subscription `v.s0`.

bcardosolopes commented on clang/lib/CIR/CodeGen/CIRGenExpr.cpp:
```diff
@@ -646,9 +646,69 @@ RValue CIRGenFunction::buildLoadOfLValue(LValue LV, SourceLocation Loc) {
   if (LV.isSimple())
     return RValue::get(buildLoadOfScalar(LV, Loc));
+
+  if (LV.isVectorElt()) {
```
Seems like this is content from the other patch? If there are dependencies between the PRs, it might be better not to put them up for review just yet, otherwise I have to go to the same thing twice and mentally account for it :)

bcardosolopes commented on clang/test/CIR/Lowering/vectype-cl.cpp:
```diff
@@ -0,0 +1,53 @@
+// RUN: %clang_cc1 -std=c++17 -triple x86_64-unknown-linux-gnu -fclangir -emit-cir %s -o %t.cir
+// RUN: cir-opt %t.cir -cir-to-llvm -o %t.mlir
```
For testing LLVM from C/C++ sources, we usually do it directly in `test/CIR/CodeGen` by adding extra `RUN` lines with `-emit-cir -emit-llvm` (we prefer to test final LLVM than LLVM dialect). Please merge this test over there. For this directory the most common is CIR sources, which are not necessary for this PR.

seven-mile replied:
Apologies about this, I copied the pattern from `vectype.cpp`. XD Will fix this.

bcardosolopes commented on clang/test/CIR/Lowering/vectype-cl.cpp:
```diff
@@ -0,0 +1,53 @@
+// RUN: %clang_cc1 -std=c++17 -triple x86_64-unknown-linux-gnu -fclangir -emit-cir %s -o %t.cir
```
Nothing opencl specific that we should use here? If we are testing stuff against regular linux do we need to mention opencl at all?

seven-mile replied:
Sorry for the confusing description on the top! In clang, the OpenCL vectors are "implemented" by extended vectors (`__attribute__((ext_vector_type(4)))`), which supports extra syntax like swizzles. Extended vectors are a "one-for-all" implementation of OpenCL vectors, Apple Metal vectors, etc. We don't support OpenCL langOpts or SPIR-V target currently, therefore x86_64 is intentionally used as a secondary option. If we don't touch the backend, then all code paths are the same. You remind me that this file is actually a test for `ext_vector_type`😂.

If it looks too strange, we may defer this PR until the initial support for OpenCL lands.

bcardosolopes replied:
Right, exactly because it's not opencl only that I'm asking that the support is added in terms of `ext_vector_type` instead! It doesn't look strange, we just need some renaming :)

bcardosolopes commented on clang/test/CIR/CodeGen/vectype-cl.cpp:
```diff
@@ -0,0 +1,202 @@
+typedef long vl2 __attribute__((ext_vector_type(2)));
+typedef unsigned short vus2 __attribute__((ext_vector_type(2)));
+
+void vector_int_test(int x) {
```
Add a check line for the function

bcardosolopes commented on clang/test/CIR/CodeGen/vectype-cl.cpp:
```diff
@@ -0,0 +1,202 @@
+  // CHECK: %{{[0-9]+}} = cir.vec.shuffle.dynamic %{{[0-9]+}} : !cir.vector<!s32i x 4>, %{{[0-9]+}} : !cir.vector<!s32i x 4>
+}
+
+void vector_double_test(int x, double y) {
```
Same here

bcardosolopes commented on clang/lib/CIR/CodeGen/CIRGenExpr.cpp:
```diff
@@ -653,9 +653,62 @@ RValue CIRGenFunction::buildLoadOfLValue(LValue LV, SourceLocation Loc) {
 }
 
+int64_t CIRGenFunction::getAccessedFieldNo(unsigned int idx,
+                                           const mlir::Attribute elts) {
```
Given the uses I've seen elsewhere, `elts` param could just be a ArrayAttr

bcardosolopes commented on clang/lib/CIR/CodeGen/CIRGenExpr.cpp:
```diff
@@ -653,9 +653,62 @@ RValue CIRGenFunction::buildLoadOfLValue(LValue LV, SourceLocation Loc) {
+  assert(elts.isa<mlir::ArrayAttr>());
+
+  auto elt = elts.cast<mlir::ArrayAttr>()[idx];
+  assert(elt.isa<mlir::IntegerAttr>());
```
If you are doing a `isa` followed by `cast`, better just do `dyn_cast` and assert if it's null.

bcardosolopes commented on clang/lib/CIR/CodeGen/CIRGenExpr.cpp:
```diff
@@ -653,9 +653,62 @@ RValue CIRGenFunction::buildLoadOfLValue(LValue LV, SourceLocation Loc) {
+  // HLSL allows treating scalars as one-element vectors. Converting the scalar
+  // IR value to a vector here allows the rest of codegen to behave as normal.
+  if (getLangOpts().HLSL && !Vec.getType().isa<mlir::cir::VectorType>()) {
+    auto DstTy =
```
Paths that are not tested should be `llvm_unreachable("NYI");`. Unless you are adding HLSL tests, please remove the code within this if statement in favor of `llvm_unreachable("NYI");`. Please do that for all other places where you didn't add tests to cover those paths.

bcardosolopes commented on clang/lib/CIR/CodeGen/CIRGenExpr.cpp:
```diff
@@ -653,9 +653,62 @@ RValue CIRGenFunction::buildLoadOfLValue(LValue LV, SourceLocation Loc) {
+  // a single element.  Just codegen as an extractelement.
+  const auto *ExprVT = LV.getType()->getAs<clang::VectorType>();
+  if (!ExprVT) {
+    int64_t InIdx = getAccessedFieldNo(0, Elts);
```
Is this path tested?

bcardosolopes replied:
Was this also tested?

seven-mile replied:
Yes, I think it's tested by `int one_elem_load = a.s2;`.

bcardosolopes commented on clang/lib/CIR/CodeGen/CIRGenExpr.cpp:
```diff
@@ -674,6 +727,84 @@ RValue CIRGenFunction::buildLoadOfBitfieldLValue(LValue LV,
 
+void CIRGenFunction::buildStoreThroughExtVectorComponentLValue(RValue Src,
+                                                               LValue Dst) {
+  mlir::Location loc = Dst.getExtVectorPointer().getLoc();
```
Same for this function, please make sure you add testcases or `llvm_unreachable("NYI");` for paths that are not tested.

bcardosolopes commented on clang/lib/CIR/CodeGen/CIRGenExpr.cpp:
```diff
@@ -965,6 +1100,70 @@ CIRGenFunction::buildPointerToDataMemberBinaryExpr(const BinaryOperator *E) {
+LValue
+CIRGenFunction::buildExtVectorElementExpr(const ExtVectorElementExpr *E) {
+  // Emit the base vector as an l-value.
+  LValue Base;
```
Same for this function, please make sure you add testcases or `llvm_unreachable("NYI");` for paths that are not tested.

bcardosolopes commented on clang/lib/CIR/CodeGen/CIRGenExpr.cpp:
```diff
@@ -965,6 +1100,70 @@ CIRGenFunction::buildPointerToDataMemberBinaryExpr(const BinaryOperator *E) {
+    // If it is a pointer to a vector, emit the address and form an lvalue with
+    // it.
+    LValueBaseInfo BaseInfo;
+    // TODO(cir): Support TBAA
```
Please also add `assert(!UnimplementedFeature::tbaa());`

bcardosolopes commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -2203,13 +2207,6 @@ mlir::Value ScalarExprEmitter::VisitAbstractConditionalOperator(
-  // OpenCL: If the condition is a vector, we can treat this condition like
-  // the select function.
-  if ((CGF.getLangOpts().OpenCL && condExpr->getType()->isVectorType()) ||
-      condExpr->getType()->isExtVectorType()) {
```
Why are you removing this check? I don't think you have testcases to back this up and doing this will probably lead to some mis-compilation that is going hard to track later. There's specific handling that needs to be done for `VisitAbstractConditionalOperator` in face of these conditions.

seven-mile replied:
`vector_ternary_lowering` is on this path. This was mentioned in the PR description (I'll update it to a brief version though):

> [A special method](https://github.com/llvm/clangir/blob/3da10fafac66ff125fb59c602e41ad4b4f5cb382/clang/lib/CodeGen/CGExprScalar.cpp#L5165) is used for the original CodeGen for ternary operator of OpenCL vectors. I don't know why. Since we already have high-level cir.vec.ternary, I skipped this part. And the lowering still works well. Maybe it's about the SPIR-V target.

Probably I'll also comment the corresponding part in the PR review ahead of time : )

bcardosolopes replied:
> Since we already have high-level cir.vec.ternary, I skipped this part

Unfortunately that's not a good rationale and not how we usually go about these things in CIR - if this leads to a miscompile, it's going to be really hard to pinpoint the issue, and might take a lot of investigation time. If something was implemented that is shortcuting here, it should also be fixed to maintain the skeleton the much pristine as possible. It's possible you might find out that you need to do another PR to fix that problem before you can land this one.

Alternatively, if you are deleting this code so one of your testcases may pass, you can then choose to not add that testcase and address the whole issue together in another PR.

bcardosolopes commented on clang/lib/CIR/CodeGen/CIRGenValue.h:
```diff
@@ -316,6 +317,20 @@ class LValue {
+    assert(isExtVectorElt());
+    return V;
+  }
+  mlir::Attribute getExtVectorElts() const {
```
From overall usage in this PR, looks like this could return an `ArrayAttr` instead.

bcardosolopes reviewed (changes requested):
Way to go on the skeleton work and thanks for fixing the testcase, more comments inline.

seven-mile commented:
Test cases improved. Tell me if I missed something! ; )

seven-mile commented on clang/lib/CIR/CodeGen/CIRGenExpr.cpp:
```diff
@@ -979,6 +1103,71 @@ CIRGenFunction::buildPointerToDataMemberBinaryExpr(const BinaryOperator *E) {
+    return LValue::MakeExtVectorElt(Base.getAddress(), elts, type,
+                                    Base.getBaseInfo());
+  }
+  assert(Base.isExtVectorElt() && "Can only subscript lvalue vec elts here!");
+
+  mlir::ArrayAttr baseElts = Base.getExtVectorElts();
+
+  // Composite the two indices
+  SmallVector<int64_t, 4> attrElts;
+  for (uint32_t i : indices) {
+    attrElts.push_back(getAccessedFieldNo(i, baseElts));
+  }
+  auto elts = builder.getI64ArrayAttr(attrElts);
+
+  return LValue::MakeExtVectorElt(Base.getExtVectorAddress(), elts, type,
+                                  Base.getBaseInfo());
```
This is covered by `int r = v.xy.x;`

seven-mile commented on clang/lib/CIR/CodeGen/CIRGenExpr.cpp:
```diff
@@ -674,6 +720,80 @@ RValue CIRGenFunction::buildLoadOfBitfieldLValue(LValue LV,
+
+      Vec = builder.createVecShuffle(loc, SrcVal, Mask);
+    } else if (NumDstElts > NumSrcElts) {
+      // Extended the source vector to the same length and then shuffle it
```
This is covered by `a.lo = b;`, in which a is `int4` and b is `int2`.

seven-mile commented on clang/lib/CIR/CodeGen/CIRGenExpr.cpp:
```diff
@@ -674,6 +720,80 @@ RValue CIRGenFunction::buildLoadOfBitfieldLValue(LValue LV,
+      // When the vector size is odd and .odd or .hi is used, the last element
+      // of the Elts constant array will be one past the size of the vector.
+      // Ignore the last element here, if it is greater than the mask size.
+      if (getAccessedFieldNo(NumSrcElts - 1, Elts) == Mask.size())
```
This is covered by `c.hi = a;`, where c is `int7` and a is `int4`.
Note that int3 is the simplest case, but blocked by [other reasons](https://github.com/llvm/clangir/blob/4ffa090986f3c98875560424ccacd9e0d1589a7a/clang/lib/CIR/CodeGen/CIRGenExpr.cpp#L2576). I'll resolve it later.

jopperm commented on clang/test/CIR/CodeGen/vectype-ext.cpp:
```diff
@@ -0,0 +1,355 @@
+  // LLVM-NEXT: store i32 %[[#EXTRACT1]], ptr %{{[0-9]+}}, align 4
+
+
+  // OpenCL C Specification 6.3.7. Vector Components
+  // The suffixes .lo (or .even) and .hi (or .odd) for a 3-component vector type
+  // operate as if the 3-component vector type is a 4-component vector type with
+  // the value in the w component undefined.
+  a = c.hi;
```
Unsure about this test, as `int7` wouldn't be a valid OpenCL vector type. Maybe you should fix the 3-element vector type support first?

jopperm reviewed (commented):
(The free-function nits are in reference to https://mlir.llvm.org/deprecation/; please check all occurrences in this PR.)

seven-mile commented:
Nice catch! I appreciate it. Updated.

Removed test cases for `int7` and the code path to support odd-component vectors. Will submit the patch for 3-component vectors and the removed part afterward.

bcardosolopes reviewed (approved):
LGTM after one minor pending inline question! Thanks!

bcardosolopes commented on clang/test/CIR/CodeGen/vectype-ext.cpp:
```diff
@@ -0,0 +1,316 @@
+
+// CIR: cir.func {{@.*vector_int_test.*}}
+// LLVM: define void {{@.*vector_int_test.*}}
+void vector_int_test(int x) {
```
Where are the LLVM checks for this and next function?

seven-mile replied:
Updated.

bcardosolopes commented on clang/test/CIR/CodeGen/vectype-ext.cpp:
```diff
@@ -0,0 +1,454 @@
+  // CIR: %{{[0-9]+}} = cir.vec.create(%{{[0-9]+}}, %{{[0-9]+}}, %{{[0-9]+}}, %{{[0-9]+}} : !s32i, !s32i, !s32i, !s32i) : !cir.vector<!s32i x 4>
+  // LLVM:      %[[#X1:]] = load i32, ptr %{{[0-9]+}}, align 4
+  // LLVM-NEXT: %[[#X2:]] = load i32, ptr %{{[0-9]+}}, align 4
+  // LLVM-NEXT: %[[#SUM:]] = add i32 %[[#X2]], 1
```
If we look at how LLVM currently codegen for this, we get a `add nsw` instead of a plain `add`, see https://godbolt.org/z/xdevh38r8. There are probably other differences too.

I believe its not related to this specific PR, but to set expectations: whenever LLVM checks are written, that should not be done blindly, we should actually check it against what traditional (OG) codegen currently emits. If there are semantic differences, they ideally should be fixed as part of the PR, but if not, we should create an issue so someone else could take a look later.

I'll create an issue for this one, can you please go through the rest of the file and see if you can spot more differences and create issues for them? Thanks in advance!

bcardosolopes replied:
Done in https://github.com/llvm/clangir/issues/664

seven-mile replied:
Many thanks for finding out this diff! I actually went through the final LLVM IR and compared them, but missed this overflow flag quite carelessly.

After a double check, I found the default index types used by two paths are different: `i64` for CIR and `i32` for Clang. But if user is using other types(like `vec[x: i32 or i64]`), in both paths, they will be preserved as is. This PR chose i64 because the previous works from @dkolsen-pgi have already done it. Since they are all self-consistent, I don't think it's a problem.

Nothing else special.

bcardosolopes replied:
Sounds good to me, thanks for double checking the rest!
